cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -246,6 +279,75 @@ public void run() {
     }
   }
 
+  private TransactionContext getTransactionContextFromAction(WorkflowAction action) {
+    WorkflowContext context;
+    if (action instanceof AbstractWorkflowAction) {
+      context = ((AbstractWorkflowAction) action).getContext();
+    } else {
+      context = ((ProgramWorkflowAction) action).getWorkflowContext();
+    }
+    return getTransactionContextFromContext(context);
+  }
+
+  private TransactionContext getTransactionContextFromContext(WorkflowContext context) {
+    TransactionContext transactionContext = new TransactionContext(txClient);
+    Collection<Dataset> datasets = ((BasicWorkflowContext) context).getDatasets().values();
+    for (Dataset ds : datasets) {
+      transactionContext.addTransactionAware((TransactionAware) ds);
+    }
+    return transactionContext;
+  }
+
+  private void runInTransaction(WorkflowAction action) {
+    TransactionContext transactionContext = getTransactionContextFromAction(action);
+    try {
+      transactionContext.start();
+      action.run();
+      transactionContext.finish();
+    } catch (TransactionFailureException e) {
+      LOG.error("Transaction failed to run: " + e.getMessage());
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private void destroyInTransaction(WorkflowAction action, WorkflowActionSpecification spec) {
+    TransactionContext transactionContext = getTransactionContextFromAction(action);
+    try {
+      transactionContext.start();
+      destroy(spec, action);
+      transactionContext.finish();
+    } catch (TransactionFailureException e) {
+      LOG.error("Transaction failed to destroy: " + e.getMessage());
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private void initializeInTransaction(WorkflowAction action, WorkflowContext workflowContext) throws Exception {
+    TransactionContext transactionContext = getTransactionContextFromContext(workflowContext);
+    try {
+      transactionContext.start();
+      action.initialize(workflowContext);
+      transactionContext.finish();
+    } catch (TransactionFailureException e) {
+      LOG.error("Transaction failed to initialize: " + e.getMessage());
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private boolean applyPredicateInTransaction(Predicate predicate, WorkflowContext context) {
```

Comment:
Lets not do it in this PR. we can create another PR for fixing predicates and unit test case over there.